File: runlens/ingest.py
"""FIT ingestion: binary FIT file -> normalized plain-dict activity.

Pure functions, no DB access. No downstream stage may reach back into raw
FIT after this module: everything the pipeline needs is in the activity
dict returned by parse_fit().

Activity dict shape (all timestamps are UTC datetimes, per verified FIT
structure — convert to local only at display time):

    {
      'file_id':       {'serial_number', 'time_created', 'manufacturer', 'product'},
      'session':       {...session summary fields...},
      'laps':          [{...}, ...],          # ordered by start_time
      'events':        [{...}, ...],          # timer + recovery_hr events
      'records':       [{...}, ...],          # 1 Hz samples, ordered
      'splits':        [{...}, ...],          # Garmin auto-splits
      'workout':       {'wkt_name': ...} | None,
      'workout_steps': [{...}, ...],          # ordered by message_index
      'time_in_zone':  [{...}, ...],
    }
"""

from __future__ import annotations

import logging

log = logging.getLogger(__name__)

# MVP record fields (spec 2.3). Other fields exist in the files but are
# deliberately not carried forward.
_RECORD_FIELDS = ("timestamp", "distance", "enhanced_speed", "heart_rate", "cadence")

_SESSION_FIELDS = (
    "start_time", "timestamp", "sport", "sub_sport",
    "total_distance", "total_timer_time", "total_elapsed_time",
    "avg_speed", "enhanced_avg_speed", "avg_heart_rate", "max_heart_rate",
    "total_ascent", "total_descent", "num_laps",
)

_LAP_FIELDS = (
    "start_time", "timestamp", "total_distance",
    "total_timer_time", "total_elapsed_time",
    "avg_speed", "enhanced_avg_speed", "avg_heart_rate", "max_heart_rate",
    "lap_trigger", "wkt_step_index", "intensity", "message_index",
)

_SPLIT_FIELDS = (
    "split_type", "start_time", "end_time",
    "total_timer_time", "total_elapsed_time", "total_distance", "avg_speed",
)

_EVENT_FIELDS = ("timestamp", "event", "event_type", "data")

_WORKOUT_STEP_FIELDS = (
    "message_index", "duration_type", "duration_distance", "duration_time",
    "duration_step", "repeat_steps", "intensity",
    "target_type", "custom_target_speed_low", "custom_target_speed_high",
    "wkt_step_name", "notes",
)

_FILE_ID_FIELDS = ("serial_number", "time_created", "manufacturer", "product", "type")

_TIME_IN_ZONE_FIELDS = ("timestamp", "reference_mesg", "time_in_hr_zone", "hr_zone_high_boundary")


def empty_activity() -> dict:
    return {
        "file_id": {},
        "session": {},
        "laps": [],
        "events": [],
        "records": [],
        "splits": [],
        "workout": None,
        "workout_steps": [],
        "time_in_zone": [],
    }


def _extract(frame, fields) -> dict:
    out = {}
    for name in fields:
        try:
            out[name] = frame.get_value(name, fallback=None)
        except Exception:  # unknown/corrupt field: skip, never crash (spec 5.3)
            out[name] = None
    return out


def parse_fit(source) -> dict:
    """Parse a FIT file (path, bytes or file-like) into an activity dict.

    Unknown or corrupt messages are skipped and logged, never fatal.
    """
    import fitdecode

    activity = empty_activity()

    with fitdecode.FitReader(source, check_crc=fitdecode.CrcCheck.WARN) as reader:
        for frame in reader:
            if not isinstance(frame, fitdecode.FitDataMessage):
                continue
            name = frame.name
            if name.startswith("unknown"):
                continue  # many unknown_* messages exist; ignore (spec 2.2)
            try:
                _dispatch(activity, name, frame)
            except Exception:
                log.warning("skipping unreadable %s message", name, exc_info=True)

    _finalize(activity)
    return activity


def _dispatch(activity: dict, name: str, frame) -> None:
    if name == "record":
        rec = _extract(frame, _RECORD_FIELDS)
        if rec.get("timestamp") is not None:
            activity["records"].append(rec)
    elif name == "lap":
        activity["laps"].append(_extract(frame, _LAP_FIELDS))
    elif name == "event":
        activity["events"].append(_extract(frame, _EVENT_FIELDS))
    elif name == "split":
        activity["splits"].append(_extract(frame, _SPLIT_FIELDS))
    elif name == "session":
        activity["session"] = _extract(frame, _SESSION_FIELDS)
    elif name == "file_id":
        activity["file_id"] = _extract(frame, _FILE_ID_FIELDS)
    elif name == "workout":
        activity["workout"] = {"wkt_name": frame.get_value("wkt_name", fallback=None)}
    elif name == "workout_step":
        activity["workout_steps"].append(_extract(frame, _WORKOUT_STEP_FIELDS))
    elif name == "time_in_zone":
        activity["time_in_zone"].append(_extract(frame, _TIME_IN_ZONE_FIELDS))
    # everything else (hrv, gps_metadata, user_profile, zones_target, ...)
    # is present in the files but unused by the MVP pipeline.


def _finalize(activity: dict) -> None:
    activity["records"].sort(key=lambda r: r["timestamp"])
    activity["laps"].sort(key=lambda l: (l.get("start_time") or _MIN_KEY,))
    activity["workout_steps"].sort(key=lambda s: s.get("message_index") or 0)
    splits = activity["splits"]
    if splits and all(s.get("start_time") is not None for s in splits):
        splits.sort(key=lambda s: s["start_time"])


class _MinKey:
    def __lt__(self, other):
        return True

    def __gt__(self, other):
        return False


_MIN_KEY = _MinKey()


def dedup_key(activity: dict) -> str:
    """Idempotent-upload identity: file_id.serial_number + time_created."""
    fid = activity.get("file_id") or {}
    return f"{fid.get('serial_number')}:{fid.get('time_created')}"


def timer_pauses(activity: dict) -> list[tuple]:
    """(stop_time, restart_time) pairs from timer events.

    Detected from event messages (event='timer', stop_all/start); callers
    can cross-check against total_elapsed_time > total_timer_time.
    """
    pauses = []
    stop_at = None
    for ev in activity["events"]:
        if ev.get("event") != "timer" or ev.get("timestamp") is None:
            continue
        et = ev.get("event_type")
        if et in ("stop_all", "stop") and stop_at is None:
            stop_at = ev["timestamp"]
        elif et in ("start",) and stop_at is not None:
            pauses.append((stop_at, ev["timestamp"]))
            stop_at = None
    return pauses

File: docs/js/ingest.js
// Ingestion: raw FIT messages -> normalized activity object.
// Mirrors runlens/ingest.py exactly (Running.MD §3). Pure, no storage.

import { parseFit, decodeEnums } from './fit.js';

export function emptyActivity() {
  return { file_id: {}, session: {}, laps: [], events: [], records: [],
           splits: [], workout: null, workout_steps: [], time_in_zone: [] };
}

/** Parse a FIT ArrayBuffer into the normalized activity dict. */
export function parseActivity(buffer) {
  const msgs = decodeEnums(parseFit(buffer));
  const a = emptyActivity();

  a.file_id = (msgs.file_id && msgs.file_id[0]) || {};
  a.session = (msgs.session && msgs.session[0]) || {};
  a.events = (msgs.event || []).map(e => ({
    timestamp: e.timestamp ?? null, event: e.event ?? null,
    event_type: e.event_type ?? null, data: e.data ?? null }));
  a.records = (msgs.record || []).filter(r => r.timestamp).map(r => ({
    timestamp: r.timestamp,
    distance: r.distance ?? null,
    enhanced_speed: r.enhanced_speed ?? r.speed ?? null,
    heart_rate: r.heart_rate ?? null,
    cadence: r.cadence ?? null,
    power: r.power ?? null,
    temperature: r.temperature ?? null }));
  a.laps = (msgs.lap || []).map(l => ({
    start_time: l.start_time ?? null, timestamp: l.timestamp ?? null,
    total_distance: l.total_distance ?? null,
    total_timer_time: l.total_timer_time ?? null,
    total_elapsed_time: l.total_elapsed_time ?? null,
    avg_speed: l.avg_speed ?? null,
    enhanced_avg_speed: l.enhanced_avg_speed ?? l.avg_speed ?? null,
    avg_heart_rate: l.avg_heart_rate ?? null,
    max_heart_rate: l.max_heart_rate ?? null,
    lap_trigger: l.lap_trigger ?? null,
    wkt_step_index: l.wkt_step_index ?? null,
    intensity: l.intensity ?? null,
    message_index: l.message_index ?? null }));
  a.splits = (msgs.split || []).map(s => ({
    split_type: s.split_type ?? null,
    start_time: s.start_time ?? null, end_time: s.end_time ?? null,
    total_timer_time: s.total_timer_time ?? null,
    total_elapsed_time: s.total_elapsed_time ?? null,
    total_distance: s.total_distance ?? null,
    avg_speed: s.avg_speed ?? null }));
  if (msgs.workout && msgs.workout.length) {
    a.workout = { wkt_name: msgs.workout[0].wkt_name ?? null };
  }
  a.workout_steps = (msgs.workout_step || []).map(expandWorkoutStep);

  finalize(a);
  return a;
}

// FIT stores workout-step duration/target as raw values whose meaning
// depends on duration_type/target_type (dynamic subfields). Expand to
// the named fields the engine reads.
function expandWorkoutStep(s) {
  const out = {
    message_index: s.message_index ?? null,
    duration_type: s.duration_type ?? null,
    duration_distance: null, duration_time: null,
    duration_step: null, repeat_steps: null,
    intensity: s.intensity ?? null,
    target_type: s.target_type ?? null,
    custom_target_speed_low: null, custom_target_speed_high: null,
    wkt_step_name: s.wkt_step_name ?? null, notes: null };
  const v = s.duration_value;
  if (v !== null && v !== undefined) {
    if (out.duration_type === 'time') out.duration_time = v / 1000;
    else if (out.duration_type === 'distance') out.duration_distance = v / 100;
    else if (String(out.duration_type).startsWith('repeat')) {
      out.duration_step = v;
      out.repeat_steps = s.target_value ?? null;
    }
  }
  if (out.target_type === 'speed') {
    if (s.custom_target_value_low) out.custom_target_speed_low = s.custom_target_value_low / 1000;
    if (s.custom_target_value_high) out.custom_target_speed_high = s.custom_target_value_high / 1000;
  }
  return out;
}

function finalize(a) {
  a.records.sort((x, y) => x.timestamp - y.timestamp);
  a.laps.sort((x, y) => (x.start_time?.getTime() ?? -Infinity) -
                        (y.start_time?.getTime() ?? -Infinity));
  a.workout_steps.sort((x, y) => (x.message_index ?? 0) - (y.message_index ?? 0));
  if (a.splits.length && a.splits.every(s => s.start_time)) {
    a.splits.sort((x, y) => x.start_time - y.start_time);
  }
  // Garmin-Connect exports carry a bogus lap `timestamp` (= session start
  // on every lap; Running.MD §3.2). Derive lap end from start + elapsed.
  for (const lap of a.laps) {
    const dur = lap.total_elapsed_time ?? lap.total_timer_time;
    if (lap.start_time && dur &&
        (!lap.timestamp || lap.timestamp <= lap.start_time)) {
      lap.timestamp = new Date(lap.start_time.getTime() + dur * 1000);
    }
  }
  // Splits without end_time: derive the same way.
  for (const s of a.splits) {
    if (s.start_time && !s.end_time && (s.total_elapsed_time || s.total_timer_time)) {
      s.end_time = new Date(s.start_time.getTime() +
        (s.total_elapsed_time ?? s.total_timer_time) * 1000);
    }
  }
}

export function dedupKey(a) {
  const f = a.file_id || {};
  const t = f.time_created instanceof Date ? f.time_created.toISOString()
    : String(f.time_created);
  return `${f.serial_number}:${t}`;
}

/** (stopTime, restartTime) pairs from timer events. */
export function timerPauses(a) {
  const pauses = [];
  let stopAt = null;
  for (const ev of a.events) {
    if (ev.event !== 'timer' || !ev.timestamp) continue;
    if ((ev.event_type === 'stop_all' || ev.event_type === 'stop') && !stopAt) {
      stopAt = ev.timestamp;
    } else if (ev.event_type === 'start' && stopAt) {
      pauses.push([stopAt, ev.timestamp]);
      stopAt = null;
    }
  }
  return pauses;
}
